# Lister og løkker

### Mål
* Kunne bruke lister i Python; herunder bla.: opprette tom liste, legge objekt (data) i liste, hente ut data fra kjent plassering (indeks) i liste.
* Kunne forklare virkemåten til for-løkker
* Kunne forklare virkemåten til while-løkker
* Kjenne til de reserverte ordene break og pass
* Kunne bruke løkker og variable til å summere tall
* Kunne bruke løkker og variable til å simulere differenslikninger og tallfølger, bla. Fibonacci-følgen
* Kunne bruke løkker og variable til å lage verditabeller, herunder bruke printf og prettyprint

#### Sekundærem mål
* Kunne bruke list comprehensions

### Anbefalte forkunnskaper
* Du kan bruke variabler i Python


## Løkker

Vi ønsker ofte å gjenta et sett med kommandoer eller intruksjoner mange ganger. Da passer det
å bruke *løkker*. Python har to typer løkker; for-løkker og while-løkker.

### While-løkker
En while løkke vil gjenta et sett med kommandoer mens en logisk betingelse er sann.
Vi kan for eksempel regne ut summen av de to, tyve, hundre eller de tusen første heltallene med
et enkelt program:

In [5]:
# Calculate the sum of the N first natural numbers

N = 100
s = 0    # s = sum
k = 0    # counter

while k <= N:
    s = s + k
    k = k + 1
    
print(s)
    

5050


### Digresjon: aritmetiske rekker
Matematikeren Gauss fikk som ung skolegutt i oppgave fra læreren sin å regne ut summen av de hundre første heltallene. Gauss hadde ikke Python eller datamaskiner tilgjengelig, men han fant følgende måte å gjøre dette på.
Man skriver først opp summen:

$$ s = 1 + 2 + 3 + 4 + \cdots + 97 + 98 + 99 + 100 $$

Dermed kan man ordne den doble summer $2s$ på følgende måte:

$$
\begin{array}{l l l l l l l l l l l l l l l l l l l}
2s & = & 1 & + & 2 & + & 3 & + & 4  & + & \cdots & + & 97 & + & 98 & + & 99 & + & 100 & + &\\
   &   & 100 & + &  99 & + & 98 & + & 97 & + & \cdots & + & 4 & + & 3 & + & 2 & + & 1 & \\ \\
2s & = & 101 & + &  101 & + & 101 & + & 101 & + & \cdots & + & 101 & + & 101 & + & 101 & + & 101 & \\
\end{array}
$$


Dermed har vi at
$$
\begin{align*}
2s = 100\cdot101 \\ \\
s = \frac{100\cdot101}{2} = 5050
\end{align*}
$$

##### Oppgave 1
a) Modifiser programmet over til å regne ut summen av de 101 første oddetallene
$$ s = 1 + 3 + 5 + 7 + \cdots + 199 + 201. $$

b) 
Sjekk at programmet virker for de 10 første oddetallene ved å beregne summen for hånd eller med med en kalkulator, og sammenlign svaret med svaret fra programmet.

c) Bruk teknikken til Gauss for å beregne denne summen for hånd.

## Lister i Python
Du kan tenke på lister i Python nesten på samme måte som lister på papir eller i et regneark. Listene er
veldig fleksible, de kan inneholde omtrent hva som helst. Videre kan vi fjerne og legge til elementer i listene på en enkel og fleksibel måte. Lister i Python er ikke svært raske, derfor vil vi senere kapitler komme tilbake til andre datastrukturer som er mer effektive for store beregninger.

Bruken blir tydelig når vi ser et par eksempeler!

In [20]:
# Vi oppretter en liste med noen tall fra togangen
togangen = [2, 4, 6, 8, 10, 12, 14, 16, 18, 20]
print(togangen)

# indeksering: begynner på 0, så 1, så 2 osv.
print(togangen[0])
print(togangen[1])
print(togangen[2])

i = togangen.index(20)
print(i)

[2, 4, 6, 8, 10, 12, 14, 16, 18, 20]
2
4
6
9


Noen ganger ønsker vi å legge til objekter på slutten av listen. La oss legge til
22 i listen over tall i togangen.

In [23]:
togangen.append(22)
print(togangen)

[2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 22, 22]


Noen ganger ønsker vi å starte med en tom liste, og heller legge til elementer etter hvert.
Dette kan for eksempel være nyttig dersom elementene kommer fra brukerinput, men vil også bli
nyttig senere når vi skal gjøre beregninger.

In [6]:
handleliste = [] #Oppretter en tom liste
print(list)
handleliste.append('hvetemel')     # legger strengen 'gjær' inn i listen
print(handleliste)

handleliste.append('gjær')
print(handleliste)

handleliste.append('melk')
print(handleliste)

<class 'list'>
['hvetemel']
['hvetemel', 'gjær']
['hvetemel', 'gjær', 'melk']


Vi kan også sortere en liste:

In [7]:
handleliste.sort()
print(handleliste)

['gjær', 'hvetemel', 'melk']


Videre kan vi fjerne elementer fra en liste:

In [8]:
handleliste.remove('hvetemel')
print(handleliste)

['gjær', 'melk']


## Traversering av lister i Python

## Løkker
Det hender ofte at vi gjør nesten den samme beregningen om igjen flere ganger. 
Hvis vi klarer å gjøre dette tilstrekkelig generelt slik at kommandoen blir den samme hver gang
kan vi bruke såkalte *løkker*.

Anta nå at vi ønsker tallene i Fibonacci-følgen i en liste:
$$ 1,\; 1,\; 2,\; 3,\; 5,\; 8,\; 13,\; 21,\; 34,\; 55,\; ... $$
Som du sikkert husker, er hvert tall i følgen summen av de to foregående tallene.
Dette kan vi uttrykke ved hjelp av differenslikningen
$$
\begin{align*}
\\
f_n & = f_{n-1} + f_{n-2},\; \text{ der } \;\;n = 3, 4, 5,... \\ \\
\end{align*}
$$
og initialbetingelsene
$$
\begin{align*}
f_1 & = 1 \\
f_2 & = 1 \\
\end{align*}
$$
Å skrive mange slike tall inn i en liste blir raskt kjedelig!
Dersom du gjør slike kjedelige og repetitive ting, kan du være sikker på at det finnes en rask metode å gjøre dette på.
Du skal nå lære å bruke løkker for å gjøre dette automatisk

In [1]:
N = 20 # antall fibonacci-tall vi ønsker i følgen

f = [1, 1] #lag en liste med de to første fibonacci-tallene

# Husk at Python indekserer lister fra 0, og ikke 1 slik vi ofte gjør i matematikk og eksempelet over
for n in range(2, N):
    f_n = f[n-1] + f[n-2]
    f.append(f_n)

print(f)

[1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377, 610, 987, 1597, 2584, 4181, 6765]


#### Oppgave
a) Det er et kjent resultat i matematikken at forholdet $f_n\;/\;f_{n-1}$ mellom et fibonaccitall og det foregående fibonacci-tallet nærmer seg det gyldne snitt når vi beveger oss lengre og lengre ut i følgen.
Øk antall iterasjoner $N$ til $N=100$, og finn det gyldne snitt med en nøyaktighet på femten desimaler.

b) Kåre Kurious lurer på hva fibonaccitall nummer én million er. Han kan ikke lagre én billion tall i en liste, for datamskinen hans har ikke nok minne. Hjelp Kåre Kurious å regne ut fibonaccitall nummer 1 000 000.

c) Kåre Kurious lurer på hva fibonaccitall nummer tusen billioner er. Kåre Kurious har ikke tolmodighet til å vente på at løkkene i python skal bli ferdig med dette. 
* Skriv tusen billioner som en tierpotens.
* Bruk resultatet i oppgave a) til å gi Kåre et matematisk uttrykk for et estimat på fibonaccitall nummer tusen billioner.